python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,378 @@
+import gc
+import heapq
+import logging
+import numbers
+import os
+import shutil
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Dict, Union, Callable, Tuple, List, Any
+
+import ray
+from ray.tune.result import NODE_IP
+from ray.util import PublicAPI
+from ray.util.annotations import DeveloperAPI
+from ray.util.ml_utils.util import is_nan
+
+MAX = "max"
+MIN = "min"
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class TrackedCheckpoint:
+    """Checkpoint tracked by a checkpoint manager.
+
+    This class is used to track checkpoints generated by trainables and trainers in
+    order to add metadata (e.g. the result, or the node where it has been created)
+    and for bookkeeping purposes.
+
+    Args:
+        dir_or_data: Checkpoint directory, checkpoint data, or a future to either.
```

Comment:
These two classes serve very different purposes though. `ray.ml.Checkpoint` is a user-facing API to ship checkpoint data around and convert between different formats.

`TrackedCheckpoint` is a developer-facing API to track checkpoint metrics and storage location.

Storage location in `ray.ml.Checkpoint` is an internal construct and shouldn't usually be accessed. The only benefit we would get from wrapping around `ray.ml.Checkpoint` is easy conversion from dict to data - but this is a small and well-defined part. The main disadvantage is making it very hard to serialize `TrackedCheckpoints`, as per default path checkpoints are always serialized when shipped. We would have to override this behavior as well. Basically we would override most of the functionality in `ray.ml.Checkpoint` and only keep the dir-dict conversion and not use the other conversions. We would also have to change many internal APIs currently working with TrackedCheckpoint. This is very little benefit for a massive refactoring.

Reply:
We're now using `Checkpoint` internally to do the conversions but won't wrap it

Reply:
So I believe @Yard1 is currently looking into optimizing the serialization for large checkpoints in `ray.ml.Checkpoint`. In general, whatever serialization optimizations we have implemented here, shouldn't that be upstreamed to `ray.ml.Checkpoint` anyways? Basically, my reasoning here is that I want to avoid essentially having 2 different Checkpoint abstractions, one for users, and one for internal, but both having a similar purpose. The distinction may be well-defined for now, but over time I'm worried these can easily diverge.

Reply:
I think the main difference is that the TrackedCheckpoint is also used to track checkpoints on remote nodes that are potentially never serialized and sent to the driver. The `ray.ml.Checkpoint` doesn't know about checkpoints on remote nodes. We've discussed introducing a custom URI like `node://<node_ip>/path/on/remote/node` but came to the conclusion that this is not well defined and opens a lot of surface area for potential bugs relating to checkpoint persistence if this is a user-facing abstraction.

Another reason is that tracked checkpoints are included in the experiment checkpoints (i.e. the TrialRunner state) and thus would be doubly serialized all the time. The alternatives here are not to use the current checkpoint directory structure at all but always and only checkpoint to the trial runner state, which is not great for various reasons (can elaborate on this more if needed). The other alternative would be to special case this, but again, then there is no point of using the Checkpoint class to begin with.

Maybe as a counter question, what would we gain from using `ray.ml.Checkpoint` here instead of a a dict or string?

Reply:
Hey @amogkam, in regards to checkpoint performance, my main concern is avoiding unnecessary serialization where possible. That will be achieved by using the same syncing functions that Tune is using. With that in mind, I don't think the work I am doing, at least right now, has any impact on what's happening here.